nixos/modules/services/hardware/pid-fan-controller.nix

```diff
@@ -0,0 +1,163 @@
+      };
+      heat_pressure_srcs = lib.mkOption {
+        type = lib.types.nonEmptyListOf lib.types.str;
+        description = "heat pressure sources which are affected by the fan";
+      };
+    };
+  };
+in
+{
+  options.services.pid-fan-controller = {
+    enable = lib.mkEnableOption "Enable PID fan controller";
+    package = lib.mkPackageOption pkgs "pid-fan-controller" { };
+    settings = {
+      interval = lib.mkOption {
+        default = 500;
+        type = lib.types.int;
+        description = "Interval between controller cycles.";
+      };
+      heat_srcs = lib.mkOption {
+        type = lib.types.listOf (lib.types.submodule heatSource);
+        description = "list of heat sources";
+        example = ''
+          [
+            {
+              name = "cpu";
+              wildcard_path = "/sys/devices/pci0000:00/0000:00:18.3/hwmon/hwmon*/temp1_input";
+              PID_params = {
+                set_point = 60;
+                P = -5.0e-3;
+                I = -2.0e-3;
+                D = -6.0e-3;
+              };
+            }
+          ];
+        '';
+      };
+      fans = lib.mkOption {
+        type = lib.types.listOf (lib.types.submodule fan);
+        description = "list of fans";
+        example = ''
+          [
+            {
+              wildcard_path = "/sys/devices/platform/nct6775.2592/hwmon/hwmon*/pwm1";
+              min_pwm = 60;
+              max_pwm = 255;
+              heat_pressure_srcs = [
+                "cpu"
+                "gpu"
+              ];
+            }
+          ];
+        '';
+      };
+    };
+  };
+  config = lib.mkIf cfg.enable {
+    environment.etc."pid-fan-settings.json".text = builtins.toJSON cfg.settings;
+
+    systemd.services.pid-fan-controller = {
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "simple";
+        ExecStart = [ (lib.meta.getExe cfg.package) ];
+        ExecStopPost = [ "${lib.meta.getExe cfg.package} disable" ];
+        Restart = "always";
+        #This service needs to run as root to write to /sys.
+        #therefore it should operate with the least amount of priviledges needed
+        ProtectHome = "yes";
+        #strict is not possible as it needs /sys
+        ProtectSystem = "full";
+        ProtectProc = "invisible";
+        PrivateNetwork = "yes";
+        NoNewPrivileges = "yes";
+        MemoryDenyWriteExecute = "yes";
+        RestrictNamespaces = "~user pid net uts mnt";
+        ProtectKernelModules = "yes";
+        RestrictRealtime = "yes";
+        SystemCallFilter = "@system-service";
+        CapabilityBoundingSet = "~CAP_KILL CAP_WAKE_ALARM CAP_IPC_LOC CAP_BPF CAP_LINUX_IMMUTABLE CAP_BLOCK_SUSPEND CAP_MKNOD";
+      };
+      # restart unit if config changed
+      restartTriggers = [ config.environment.etc."pid-fan-settings.json".text ];
+    };
+    #sleep hook to restart the service as it breaks sometimes otherwise
```

Comment:
This should probably be fixed upstream.

Reply:
The issue arises from suspend writing to the `pwm_enable` file in the sysfs. The only way to fix it upstream i can think of would be to write to `pwm_enable` every time before writing to the `pwm` file which would could increase the time spent in kernel-space substantially as the I2C interface of most controllers is really slow (a write takes up to 150µs from what i could measure).

Having a sleep hook is more elegant compared to that and also prevents a large ramp-up when exiting suspend if the fans where running on max before entering suspend.

Reply:
I see. Still, as you are the upstream developer, I'd rather see these systemd services be included in your package, and added to the module with `systemd.packages`. This is a non-blocking best-practices nit though.

Reply:
In the past i shipped systemd units but i removed them when i integrated the flake as i want to properly package it. I will look into it though.